### ISSUE TITLE
Intercept "ctrl-r" / "cmd-r" in instances where work might be lost.

### ISSUE THREAD

**Issue by Pomax** [mozilla/thimble.mozilla.org#1526]
I've run into multiple instances where I think I'm in a browser and so instinctively hit "reload page" as keyboard shortcut to see the updated changes. Getting a "are you sure you want to leave this page?" using a `window.onreload` might be a good idea

**Comment by aayushsanghavi** [comment]
I tried setting up the project on my machine but I couldn't. Although I've written some javascript code in a separate file that resolves this issue (It throws an alert on pressing ctrl r before reloading the page). Can I just add the file and make a PR and then maybe one of the contributors could integrate it.

**Comment by gideonthomas** [comment]
@aayushsanghavi apologies for the delayed response.

Sure, feel free to open up a PR with your code.

We are also happy to help you with the problems you are having setting up Thimble on your local machine. Just let us know where (or in which step) you are experiencing issues and what operating system you are running.

**Comment by aayushsanghavi** [comment]
Thanks @gideonthomas for your help but I don't remember where I was stuck earlier. So I guess I'll just add a new JavaScript file to the repository. All that has to be done is to call include that file in one of the main code files. Kindly review the pull request I make.

**Comment by flukeout** [comment]
Based on the work by @aayushsanghavi in #1602 we've come to realize that in order for this to be useful, we'd have to detect when you are actually risking losing work by refreshing. The most obvious case for this is when the project you have hasn't fully synced with the remote DB.

If we can detect that case, then it might be useful to show an alert when the user tries to leave the editor. We can make that the focus of this issue. Can we reliably detect that state?

**Comment by humphd** [comment]
We deal with that case here:

https://github.com/mozilla/thimble.mozilla.org/blob/fa3dde088dc016a562520e62a9df1f5558ba8f70/public/editor/scripts/editor/js/fc/sync-state.js

Likely the real fix is to figure out why we're not doing what you expect there, since we already have `beforeunload` handlers to stop the page from closing.

**Comment by Pomax** [comment]
I'm going to close this for now, as it's become a bit of an undefined issue, and I'll have to run into this again before I can properly explain the problem again.

**Comment by ryanwarsaw** [comment]
@Pomax Are you okay with me re-opening this? It's definitely a race condition issue, not as much of an issue for those on strong & stable internet connections, but still an issue nonetheless.

**Comment by Pomax** [comment]
no, we decided to keep this closed during triage because the actual problem hasn't been experience in forever and so there is no reliable information on *what* part of it needs fixing to address user needs.

**Comment by ryanwarsaw** [comment]
@Pomax I was just able to replicate this exact issue, where I had paste a block of code into my Thimble project, then hit the reload key combo. The page refreshed and the changes disappeared.

**Comment by humphd** [comment]
@ryanwarsaw did you paste and hit refresh in less than 5 seconds?  We autosave to the browser 5s after you make a change.

The right fix for this is probably to hook into the events happening in https://github.com/mozilla/brackets/issues/607 for dirty/saved state changes.

**Comment by ryanwarsaw** [comment]
@humphd Yeah, I demonstrated the issue here: https://cl.ly/1u411T0I3c2S. 

@aayushsanghavi Had inquired about this issue earlier today, if he still wants to work on this I'll see if we can work together and knock this out, otherwise I'll bare responsibility and make sure this gets closed in a timely manner.

**Comment by aayushsanghavi** [comment]
@ryanwarsaw I would like that, we can work together on this one.

**Comment by flukeout** [comment]
@ryanwarsaw Another option would be to intercept the Command/Ctrl + R key combo and handle that case. We could either...

* Reload the current Preview window
* Show a popup that says why we blocked the reload and allow the user to disable that behaviour

**Comment by humphd** [comment]
OK, the PR for https://github.com/mozilla/brackets/issues/607 has landed.  This means we have 2 new events in Bramble that we can use in Thimble to help this issue.

I'll bring a few images over here so we can continue to develop the UI side of this.  First, consider how Google Docs works:

![dirty-saved](https://cloud.githubusercontent.com/assets/427398/23480138/db63e01e-fe94-11e6-9e8d-9a1ed309cc1e.gif)

And here is Medium's editor:

![medium-example](https://cloud.githubusercontent.com/assets/427398/23961907/c92d2934-0982-11e7-9c69-6fb44841ab14.gif)

We have 2 new events on the `bramble` object: `projectDirty` (there are unsaved changes in the editor), and `projectSaved` (all files are saved to the filesystem--not the same as synced to the server).

I'd like to see us do a few things, and probably we should split this up into multiple bugs:

* Add text to the header that says "Saving..." and "All files Saved." based on these two events.  We can fine tune the exact text and look in whichever bug fixes this.
* We should deal with both cases of a user being a) anonymous so there is no saving to the server; b) a logged in user who is saving locally *and* syncing to the server (we might want to have a 3-part message that indicates files are dirty, saved locally, saved to server.
* We should add `beforeunload` logic so we block attempts to reload or exit the tab while the project is dirty.
* We already have `bramble.saveAll()` right now that we could call in the `beforeunload` event to force the files to write now before the user bails.

I see @ryanwarsaw, @aayushsanghavi, and @Taguiar all involved in some portion of this bug.  Can you all figure out who is going to fix what, and file new bugs or do PRs related to what I've listed above?  I know @Taguiar needs more project work for my course, so if he can do it, that might be ideal for his grade.

**Comment by ryanwarsaw** [comment]
I can file a bug for your first outlined change, that seems like a good first bug for someone. Your second change is fairly relevant to the work @aayushsanghavi did previously on this, so I'll let him decide whether or not he wants to continue with this, or with his work on other bugs.

Also, do we need to do anything other than call ```bramble.saveAll()``` in the ```beforeunload``` event? You mentioned blocking it with an alert or something, if possible I'd like to avoid that so I'm wondering if you could elaborate on what we would need to do there.

**Comment by humphd** [comment]
There is no way to guarantee that data is always saved: power failure, browser crash, corrupt indexeddb, the list goes on.  All you can do is *try* to save, and report back if things don't work.

The events that @Taguiar wired into Brackets will help us, because we'll know when the project is not-saved vs. saved in the db.  In all cases where it's not-saved, we'll want to make sure a `beforeunload` event handler is present, and in cases where it's saved, we can remove it.  The `bramble.saveAll()` call isn't really necessary, but it's an extra safe-guard we could do in the `beforeunload` to try and force things to happen now-right-now.  But in the end, we can't block the tab staying open on an async event, so we always have some risk.

**Comment by humphd** [comment]
This has stalled.  I'm going to remove the Assigned label until someone wants to work on it.
